Code/GraphMol/FileParsers/CMLWriter.cpp

```diff
@@ -0,0 +1,209 @@
+    atom.put("<xmlattr>.x3", xyz_fmt % pos.x);
+    atom.put("<xmlattr>.y3", xyz_fmt % pos.y);
+    atom.put("<xmlattr>.z3", xyz_fmt % pos.z);
+  }
+
+  molecule.put("<xmlattr>.formalCharge", mol_formal_charge);
+  molecule.put("<xmlattr>.spinMultiplicity", mol_num_radical_electrons + 1u);
+
+  // bond/@id so that it can be referenced
+  // http://www.xml-cml.org/convention/molecular#bond-id
+  const auto bond_id_prefix = "b";
+  unsigned bond_id = 0u;
+
+  auto& bondArray = molecule.add("cml:bondArray", "");
+  for (auto atom_itr = rwmol.beginAtoms(), atom_itr_end = rwmol.endAtoms();
+       atom_itr != atom_itr_end; ++atom_itr) {
+    const auto& atom = *atom_itr;
+    PRECONDITION(atom, "bad atom");
+    const auto src = atom->getIdx();
+    const auto* const mol = &atom->getOwningMol();
+
+    for (auto bond_itrs = mol->getAtomBonds(atom);
+         bond_itrs.first != bond_itrs.second; ++bond_itrs.first) {
+      auto* bptr = (*mol)[*bond_itrs.first];
+      auto* nptr = bptr->getOtherAtom(atom);
+      const auto dst = nptr->getIdx();
+      if (dst < src) continue;
+
+      auto& bond = bondArray.add("cml:bond", "");
+      bond.put("<xmlattr>.atomRefs2",
+               boost::format{"%1%%2% %1%%3%"} % atom_id_prefix % src % dst);
+
+      bond.put("<xmlattr>.id",
+               boost::format{"%1%%2%"} % bond_id_prefix % bond_id++);
+
+      const auto btype = bptr->getBondType();
+      switch (btype) {
+        case Bond::SINGLE:
+          bond.put("<xmlattr>.order", 1);
+          break;
+        case Bond::DOUBLE:
+          bond.put("<xmlattr>.order", 2);
+          break;
+        case Bond::TRIPLE:
+          bond.put("<xmlattr>.order", 3);
+          break;
+        case Bond::AROMATIC:
+          bond.put("<xmlattr>.order", 'A');
+          break;
+        case Bond::DATIVEONE:
+          FALLTHROUGH;
+        case Bond::DATIVE:
+          FALLTHROUGH;
+        case Bond::DATIVEL:
+          FALLTHROUGH;
+        case Bond::DATIVER:
+          bond.put("<xmlattr>.order", 1);
+          break;
+        // XXX RDKit extension: bond orders greater than 3
+        case Bond::QUADRUPLE:
+          bond.put("<xmlattr>.order", 4);
+          break;
+        case Bond::QUINTUPLE:
+          bond.put("<xmlattr>.order", 5);
+          break;
+        case Bond::HEXTUPLE:
+          bond.put("<xmlattr>.order", 6);
+          break;
+        // XXX RDKit extension: half-integer orders
+        case Bond::ONEANDAHALF:
+          bond.put("<xmlattr>.order", "1.5");
+          break;
+        case Bond::TWOANDAHALF:
+          bond.put("<xmlattr>.order", "2.5");
+          break;
+        case Bond::THREEANDAHALF:
+          bond.put("<xmlattr>.order", "3.5");
+          break;
+        case Bond::FOURANDAHALF:
+          bond.put("<xmlattr>.order", "4.5");
+          break;
+        case Bond::FIVEANDAHALF:
+          bond.put("<xmlattr>.order", "5.5");
+          break;
+        default:
+          bond.pop_back();
```

Comment:
Shouldn't this never happen?    Would this be an error?   I might at least log this.

Reply:
An empty string is used, which represents unknown or unspecified type.